File: src/java/nz/co/searchwellington/tagging/AutoTaggingService.java
package nz.co.searchwellington.tagging;

import java.util.Set;

import org.apache.log4j.Logger;

import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.repositories.HandTaggingDAO;
import nz.co.searchwellington.repositories.UserRepository;

public class AutoTaggingService {
	
	static Logger log = Logger.getLogger(AutoTaggingService.class);
	
	private static final String AUTOTAGGER_PROFILE_NAME = "autotagger";
   
	private ImpliedTagService impliedTagService;
	private PlaceAutoTagger placeAutoTagger;
	private TagHintAutoTagger tagHintAutoTagger;
	private HandTaggingDAO handTaggingDAO;
	private UserRepository userDAO;
	
	public AutoTaggingService(ImpliedTagService impliedTagService, PlaceAutoTagger placeAutoTagger, TagHintAutoTagger tagHintAutoTagger, HandTaggingDAO handTaggingDAO, UserRepository userDAO) {
		this.impliedTagService = impliedTagService;
		this.placeAutoTagger = placeAutoTagger;
		this.tagHintAutoTagger = tagHintAutoTagger;
		this.handTaggingDAO = handTaggingDAO;
		this.userDAO = userDAO;
	}

	
	public void autotag(Resource resource) {
		User autotaggerUser = userDAO.getUserByProfileName(AUTOTAGGER_PROFILE_NAME);
		if (autotaggerUser == null) {
			log.warn("Could not find auto tagger user: " + AUTOTAGGER_PROFILE_NAME);
			return;
		}
		
		Set<Tag> suggestedTags = placeAutoTagger.suggestTags(resource);
		log.info("Suggested tags for '" + resource.getName() + "' are: " + suggestedTags.toString());
		suggestedTags.addAll(tagHintAutoTagger.suggestTags(resource));
		for (Tag tag : suggestedTags) {
			if (!impliedTagService.alreadyHasTag(resource, tag)) {
				log.info("Autotagging resource '" + resource.getName() + "' with " + tag.getName());
				handTaggingDAO.addTag(autotaggerUser, tag, resource);
			} else {
				log.info("Resource already has tag '" + tag.getName() + "'; ignoring");
			}
		}
	}

}

File: src/main/scala/nz/co/searchwellington/tagging/AutoTaggingService.scala
package nz.co.searchwellington.tagging

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.feeds.whakaoko.model.Category
import nz.co.searchwellington.model.taggingvotes.HandTagging
import nz.co.searchwellington.model.{Newsitem, Tag}
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.{ExecutionContext, Future}

@Component class AutoTaggingService @Autowired()(placeAutoTagger: PlaceAutoTagger,
                                                 tagHintAutoTagger: TagHintAutoTagger,
                                                 mongoRepository: MongoRepository)
  extends ReasonableWaits {

  private val log = LogFactory.getLog(classOf[AutoTaggingService])

  private val AUTOTAGGER_PROFILE_NAME = "autotagger"

  def autoTagsForFeedCategories(feedItemCategories: Seq[Category])(implicit ec: ExecutionContext): Future[Set[HandTagging]] = {
    tagHintAutoTagger.suggestFeedCategoryTags(feedItemCategories).flatMap( suggestedTags =>
      toHandTagging(suggestedTags, Some("RSS category")))
  }

  def autotag(resource: Newsitem)(implicit ec: ExecutionContext): Future[Set[HandTagging]] = {
    val eventualSuggestedPlaces = placeAutoTagger.suggestTags(resource)
    val eventualAutoTags = tagHintAutoTagger.suggestTags(resource)
    for {
      suggestedPlaces <- eventualSuggestedPlaces
      suggestedAutoTags <- eventualAutoTags
      suggestedTags = {
        val suggestedTags = suggestedPlaces ++ suggestedAutoTags
        log.info("Suggested tags for '" + resource.title + "' are: " + suggestedTags.map(_.id).mkString(","))
        suggestedTags
      }
      handTaggings <- toHandTagging(suggestedTags)
    } yield {
      handTaggings
    }
  }

  private def toHandTagging(suggestedTags: Set[Tag], reason: Option[String] = None)(implicit ec: ExecutionContext): Future[Set[HandTagging]] = {
    mongoRepository.getUserByProfilename(AUTOTAGGER_PROFILE_NAME).map { maybyAutotagUser =>
      maybyAutotagUser.map { autotagUser =>
        suggestedTags.map(t => HandTagging(tag = t, user = autotagUser, reason = reason))
      }.getOrElse {
        log.warn("Could not find auto tagger user: " + AUTOTAGGER_PROFILE_NAME + "; not autotagging.")
        Set.empty
      }
    }
  }

}
